akka-typed/src/main/scala/akka/typed/Impl.scala

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (C) 2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.typed
+
+import scala.reflect.ClassTag
+import akka.{ actor ⇒ a }
+import scala.concurrent.duration.Duration
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.ExecutionContextExecutor
+import akka.event.LoggingReceive
+
+/**
+ * INTERNAL API. Mapping the execution of a [[Behavior]] onto a good old untyped
+ * [[akka.actor.Actor]].
+ */
+private[typed] class ActorAdapter[T: ClassTag](_initialBehavior: () ⇒ Behavior[T]) extends akka.actor.Actor {
+  import Behavior._
+
+  val clazz = implicitly[ClassTag[T]].runtimeClass
+
+  var behavior = _initialBehavior()
+  val ctx = new ActorContextAdapter[T](context)
+
+  def receive = LoggingReceive {
+    case akka.actor.Terminated(ref)   ⇒ next(behavior.management(ctx, Terminated(ActorRef(ref))))
+    case akka.actor.ReceiveTimeout    ⇒ next(behavior.management(ctx, ReceiveTimeout))
+    case msg if clazz.isInstance(msg) ⇒ next(behavior.message(ctx, msg.asInstanceOf[T]))
+  }
+
+  private def next(b: Behavior[T]): Unit = {
+    behavior = canonicalize(ctx, b, behavior)
+    if (!isAlive(behavior)) {
+      context.stop(self)
+    }
+  }
+
+  override val supervisorStrategy = a.OneForOneStrategy() {
+    case ex ⇒
+      import Failed._
+      import akka.actor.{ SupervisorStrategy ⇒ s }
+      ctx.setFailureResponse(Failed.NoFailureResponse)
+      next(behavior.management(ctx, Failed(ex, ActorRef(sender()))))
+      ctx.getFailureResponse match {
+        case Resume  ⇒ s.Resume
+        case Restart ⇒ s.Restart
+        case Stop    ⇒ s.Stop
+        case _       ⇒ s.Escalate
+      }
+  }
+
+  override def preStart(): Unit =
+    next(behavior.management(ctx, PreStart))
+  override def preRestart(reason: Throwable, message: Option[Any]): Unit =
+    next(behavior.management(ctx, PreRestart(reason)))
```

Comment:
is the akka.actor.Actor default `preRestart` implemented elsewhere or not at all?